runtime/lua/vim/lsp/buf.lua

```diff
@@ -701,7 +688,29 @@ function M.typehierarchy(kind)
         end
       end)
     end
-  end)
+  end
+
+  local results = {} --- @type [integer, lsp.TypeHierarchyItem][]
+
+  local remaining = #clients
+
+  for _, client in ipairs(clients) do
+    local params = util.make_position_params(win, client.offset_encoding)
+    client.request(method, params, function(err, result, ctx)
+      if err then
+        vim.notify(err.message, vim.log.levels.WARN)
+      elseif result then
+        for _, item in pairs(result) do
+          results[#results + 1] = { ctx.client_id, item }
```

Comment:
On review, I think looking up the client is a bit more robust in case the client exists. In order to keep references to clients, I think we need to some active tracking into the client. AFAIK you can only tell if a client has exited by using `all_clients`.